topology.py

```diff
@@ -1,42 +1,164 @@
-        # TODO this will be placed on etcd in the future
-        self.nodes_metadata_store = nodes_metadata_store
-
-    def get_next_node_id(self, id):
-        if self.nodes_metadata_store.index(id) == len(self.nodes_pipeline) - 1:
-            return None
-        return self.nodes_pipeline[self.nodes_pipeline.index(id) + 1]
-
-    def get_prev_node_id(self, id):
-        if self.nodes_metadata_store.index(id) == 0:
-            return None
-        return self.nodes_pipeline[self.nodes_pipeline.index(id) - 1]
-
-    def append(self, id, ip_addr, port):
-        # TODO: in the future, we will have to get the node's IP address
-        # and port via etcd
-        self.nodes_metadata_store[id] = hooke_pb2.NodeInfo(
-            ip_addr=ip_addr, node_id=id, port=port
-        )
-        self.nodes_pipeline.append(id)
 
+        # self.etcd is the inbuilt database connection for node metadata
+        self.etcd_client = etcd3.client(host=etcd_host, port=etcd_port)
+
+        # self.addr is just the address for nodes - e.g. localhost
+        # assumes every node will have the same address - unsure if safe assumption
+        self.addr = addr
+
+        # use logging instead of print statements
+        logging.basicConfig(level=logging.INFO)
+        self.logger = logging.getLogger(__name__)
+
+    def get_next_node_id(self, node_id):
+        """
+        Gets the id for the next node. Returns -1 if it doesn't exist.
+        """
+        if node_id >= len(self.nodes_pipeline) - 1:
+            return -1
+        
+        try:
+            next_node = self.etcd_client.get(f"nodes/{node_id}/next_node")
+
+            if not next_node or not next_node[0]:
+                return -1
+            
+            return int(next_node[0])
+        
+        except Exception as e:
+            self.logger.error("Error getting next node id for node %s: %s", node_id, e)
+            return -1
+    
+    def get_prev_node(self, node_id):
+        """
+        Gets the id for the previous node. Returns -1 if it doesn't exist.
+        """
+        if node_id == 0:
+            return -1
+
+        try:
+            prev_node = self.etcd_client.get(f"nodes/{node_id}/prev_node")
+
+            if not prev_node or not prev_node[0]:
+                return -1
+        
+            return int(prev_node[0])
+        
+        except Exception as e:
+            self.logger.error("Error getting previous node id for node %s: %s", node_id, e)
+            return -1
+        
+    def append(self, node_id):
+        """
+        Adds a node to the topology and assigns it a port, stored in etcd. Also updates next_node for the previous node.
+        """
+        prev_node = self.nodes_pipeline[-1] if self.nodes_pipeline else -1
+        port = network.find_available_port(self.addr)
+        
+        if port == -1:
+            # TODO: handle the case where there's no open ports
+            return
+        
```

Comment:
I think there are some confusions here. So by the time append() is called. The node's process should already been binded to a port. So `find_available_port` should be used somewhere in node.py instead
What I meant in the TODO comment is that instead of having that port passed in, we could retrieve it from the metadata store.